src/runtime/format.c

```diff
@@ -126,11 +122,8 @@ void bprintf(buffer b, const char *fmt, ...)
 
 void rprintf(const char *format, ...)
 {
-    static char buf[PAGESIZE * 4];
-
-    buffer b = alloca_wrap_buffer(buf, sizeof(buf));
-    b->end = 0;
-
+    /* What's a reasonable limit here? This needs to be reentrant. */
```

Comment:
it really wouldn't be much on top after this is done to add a thread-safe heap, just for this stuff around the edges..especially since the the bottom-most layer (the pages) is safe. or a per-cpu heap with no lock might even be better